include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -1,14 +1,98 @@
-#version 430 core
+#ifndef _IRR_BUILTIN_MATERIAL_COMPILER_GLSL_RASTERIZATION_IMPL_INCLUDED_
+#define _IRR_BUILTIN_MATERIAL_COMPILER_GLSL_RASTERIZATION_IMPL_INCLUDED_
 
 #include <irr/builtin/material_compiler/glsl/common.glsl>
 
-void main()
+void instr_eval_execute(in instr_t instr, in vec3 L)
 {
-	mat2 dUV = mat2(dFdx(UV),dFdy(UV));
+	uint op = instr_getOpcode(instr);
 
-	InstanceData instData = InstData.data[InstanceIndex];
-	runTexPrefetchStream(instData.prefetch_instrStream);
-	runNormalPrecompStream(instData.nprecomp_instrStream);
+	//speculative execution
+	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	params_t params = instr_getParameters(instr, bsdf_data);
+	float bxdf_eval_scalar_part;
+	uint ndf = instr_getNDF(instr);
+	float a = getAlpha(params);
+	float a2 = a*a;
+	float ay = getAlphaV(params);
+	float ay2 = ay*ay;
 
+	if (op_hasSpecular(op))
+	{
+		if (ndf==NDF_GGX) {
+			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
+		}
+		else if (ndf==NDF_BECKMANN) {
+			bxdf_eval_scalar_part = irr_glsl_beckmann_smith_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
+		}
```

Comment:
what about aniso GGX and beckmann?

Reply:
right, i forgot backend doesnt just follow mitsuba only